src/v/kafka/data/cloud_topic_partition.cc

```diff
@@ -0,0 +1,599 @@
+    // path
+
+    // This part comprises the control plane query. The partition is a metadata
+    // storage for the cloud topic.
+    auto underlying = co_await _partition->make_reader(cfg, timeout);
+    auto placeholders = co_await model::consume_reader_to_chunked_vector(
+      std::move(underlying), model::no_timeout);
+
+    // TODO: convert to a extent_meta
+    auto extents = [ph = std::move(placeholders)]() mutable {
+        chunked_vector<experimental::cloud_topics::extent_meta> res;
+        for (auto&& batch : ph) {
+            experimental::cloud_topics::extent_meta e{
+              .base_offset = model::offset_cast(batch.base_offset()),
+              .last_offset = model::offset_cast(batch.last_offset()),
+            };
+            iobuf payload = std::move(batch).release_data();
+            iobuf_parser parser(std::move(payload));
+            auto record = model::parse_one_record_from_buffer(parser);
+            iobuf value = std::move(record).release_value();
+            auto placeholder
+              = serde::from_iobuf<experimental::cloud_topics::dl_placeholder>(
+                std::move(value));
+            e.id = placeholder.id;
+            e.first_byte_offset = placeholder.offset;
+            e.byte_range_size = placeholder.size_bytes;
+            res.push_back(e);
+        }
+        return res;
+    }();
+
+    // This part comprises the data plane query. The 'cloud_topics::api' is
+    // responsible for moving the data from the cloud storage to the local
+    // cache.
+    auto data_batches = co_await _ct_api->materialize(
+      ntp(),
+      cfg.max_bytes,
+      std::move(extents),
+      // TODO: use configurable default timeout or derive from the log reader
+      // config
+      10s);
+
+    if (!data_batches) {
+        throw std::system_error(data_batches.error());
+    }
+
+    co_return storage::translating_reader(
+      model::make_fragmented_memory_record_batch_reader(
+        std::move(data_batches.value())),
+      ot_state);
+}
+
+ss::future<std::vector<cluster::tx::tx_range>>
+cloud_topic_partition::aborted_transactions(
+  model::offset base,
+  model::offset last,
+  ss::lw_shared_ptr<const storage::offset_translator_state> ot_state) {
+    // TODO: add code path that fetches metadata from the dl_stm for L1 read
+    auto base_rp = ot_state->to_log_offset(base);
+    auto last_rp = ot_state->to_log_offset(last);
+    cloud_storage::offset_range offsets = {
+      .begin = model::offset_cast(base),
+      .end = model::offset_cast(last),
+      .begin_rp = base_rp,
+      .end_rp = last_rp,
+    };
+    co_return co_await get_aborted_transactions_local(*_partition, offsets);
+}
+
+ss::future<std::optional<storage::timequery_result>>
+cloud_topic_partition::timequery(storage::timequery_config cfg) {
+    // cluster::partition::timequery returns a result in Kafka offsets,
+    // no further offset translation is required here.
```

Comment:
Well the timestamps are different between the placeholder batches and the input batches. Should we fix?